File: vol2/ch25/LeetCode380.cpp
class RandomizedSet{
   vector<int> nums;                   					//存放全部元素
        unordered_map<int,int> hmap;               //<整数,索引>哈希映射
    public:
    RandomizedSet() { }
        bool insert(int val) {                              //插入val
        if(hmap.find(val)==hmap.end()) {          //元素不存在时插入
            nums.push_back(val);                //在nums末尾插入val
            hmap[val]=nums.size()-1;            //val的索引为n-1
            return true;
        }
        return false;
    }
    bool remove(int val) {                              //删除val
        if(hmap.find(val)!=hmap.end()) {          //找到key
            int i=hmap[val];                           //找到val的索引i
            int lastk=nums[nums.size()-1];              //求末尾关键字lastk
            nums[i]=lastk;                          //将末尾关键字移到i索引处
            hmap[lastk]=i;						//重置hmap[lastk]
            hmap.erase(val);                           //删除val关键字
            nums.pop_back();                //将nums末尾元素删除
            return true;
        }
        return false;
    }
    int getRandom() {                       //随机返回一个key
        int j=rand()%nums.size();                //返回0~n-1的随机数
        return nums[j];
    }
};

File: vol2/ch25/LeetCode380.py
class RandomizedSet:
    def __init__(self):
        self.nums=[]                  #存放全部元素
        self.hmap={}               #<整数,索引>哈希表

    def insert(self, val: int) -> bool:
        if val not in self.hmap:     #元素不存在时插入
            self.nums.append(val)               #在nums末尾插入val
            self.hmap[val]=len(self.nums)-1       #val的索引为n-1
            return True
        return False

    def remove(self, val: int) -> bool:
        if val in self.hmap:          #找到key
            i=self.hmap[val]                          #找到val的索引i
            lastk=self.nums[len(self.nums)-1]              #求末尾关键字lastk
            self.nums[i]=lastk                      #将末尾关键字移到i索引处
            self.hmap[lastk]=i					#重置hmap[lastk]
            del self.hmap[val]                     #删除val关键字
            self.nums.pop()                    #将nums末尾元素删除
            return True
        return False

    def getRandom(self) -> int:
        j=random.randrange(0,len(self.nums),1)             #返回0~n-1的随机数
        return self.nums[j]
